// @atlaskit/adf-schema/dist/cjs/schema/jira-schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = makeSchema;
exports.isSchemaWithAdvancedTextFormattingMarks = isSchemaWithAdvancedTextFormattingMarks;
exports.isSchemaWithBlockQuotes = isSchemaWithBlockQuotes;
exports.isSchemaWithCodeBlock = isSchemaWithCodeBlock;
exports.isSchemaWithEmojis = isSchemaWithEmojis;
exports.isSchemaWithLinks = isSchemaWithLinks;
exports.isSchemaWithLists = isSchemaWithLists;
exports.isSchemaWithMedia = isSchemaWithMedia;
exports.isSchemaWithMentions = isSchemaWithMentions;
exports.isSchemaWithSubSupMark = isSchemaWithSubSupMark;
exports.isSchemaWithTables = isSchemaWithTables;
exports.isSchemaWithTextColor = isSchemaWithTextColor;

var _createSchema = require("./create-schema");

function makeSchema(config) {
  var nodes = ['doc', 'paragraph', 'text', 'hardBreak', 'heading', 'rule'];
  var marks = ['strong', 'em', 'underline', 'typeAheadQuery', 'unsupportedMark', 'unsupportedNodeAttribute'];

  if (config.allowLinks) {
    marks.push('link');
  }

  if (config.allowLists) {
    nodes.push('orderedList', 'bulletList', 'listItem');
  }

  if (config.allowMentions) {
    nodes.push('mention');
    marks.push('mentionQuery');
  }

  if (config.allowEmojis) {
    nodes.push('emoji');
  }

  if (config.allowAdvancedTextFormatting) {
    marks.push('strike', 'code');
  }

  if (config.allowSubSup) {
    marks.push('subsup');
  }

  if (config.allowCodeBlock) {
    nodes.push('codeBlock');
  }

  if (config.allowBlockQuote) {
    nodes.push('blockquote');
  }

  if (config.allowMedia) {
    nodes.push('mediaGroup', 'mediaSingle', 'media', 'caption', 'mediaInline');
  }

  if (config.allowTextColor) {
    marks.push('textColor');
  }

  if (config.allowTables) {
    nodes.push('table', 'tableCell', 'tableHeader', 'tableRow');
  }

  return (0, _createSchema.createSchema)({
    nodes: nodes,
    marks: marks
  });
}

function isSchemaWithLists(schema) {
  return !!schema.nodes.bulletList;
}

function isSchemaWithMentions(schema) {
  return !!schema.nodes.mention;
}

function isSchemaWithEmojis(schema) {
  return !!schema.nodes.emoji;
}

function isSchemaWithLinks(schema) {
  return !!schema.marks.link;
}

function isSchemaWithAdvancedTextFormattingMarks(schema) {
  return !!schema.marks.code && !!schema.marks.strike;
}

function isSchemaWithSubSupMark(schema) {
  return !!schema.marks.subsup;
}

function isSchemaWithCodeBlock(schema) {
  return !!schema.nodes.codeBlock;
}

function isSchemaWithBlockQuotes(schema) {
  return !!schema.nodes.blockquote;
}

function isSchemaWithMedia(schema) {
  return !!schema.nodes.mediaGroup && !!schema.nodes.media && !!schema.nodes.mediaInline;
}

function isSchemaWithTextColor(schema) {
  return !!schema.marks.textColor;
}

function isSchemaWithTables(schema) {
  return !!schema.nodes.table && !!schema.nodes.tableCell && !!schema.nodes.tableHeader && !!schema.nodes.tableRow;
}
