// @atlaskit/adf-schema/dist/esm/next-schema/nodes/inlineCard.js
import { adfNode } from '@atlaskit/adf-schema-generator';
import { annotation } from '../marks/annotation';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
export var inlineCard = adfNode('inlineCard').define({
  inline: true,
  selectable: true,
  draggable: true,
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    anyOf: [{
      url: {
        type: 'string',
        default: null,
        validatorFn: 'safeUrl'
      }
    }, {
      data: {
        type: 'object',
        default: null
      }
    }]
  },
  stage0: {
    marks: [annotation, unsupportedNodeAttribute, unsupportedMark]
  },
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.marks': {
        reason: '@DSLCompatibilityException - Annotation mark is in variant, but not full schema. However this has been added for compatibility with current validator.',
        value: {
          items: ['annotation'],
          optional: true,
          type: 'array'
        }
      }
    }
  }
}).variant('with_annotation', {
  marks: [annotation, unsupportedMark, unsupportedNodeAttribute]
});
