// @atlaskit/adf-schema/dist/esm/next-schema/nodes/tableHeader.js
import { adfNode } from '@atlaskit/adf-schema-generator';
import { tableHeaderContentPseudoGroup } from '../groups/tableCellContentPseudoGroup';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
export var tableHeader = adfNode('tableHeader').define({
  isolating: true,
  selectable: false,
  tableRole: 'header_cell',
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    colspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    rowspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    colwidth: {
      type: 'array',
      items: {
        type: 'number'
      },
      default: null,
      optional: true
    },
    background: {
      type: 'string',
      default: null,
      optional: true
    }
  },
  content: [tableHeaderContentPseudoGroup],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      required: {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: ['content']
      },
      'props.content': {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: 'tableCell_content'
      }
    }
  }
});
