// @atlaskit/adf-schema/dist/cjs/steps/table/sort-column.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tableSortingStepType = exports.TableSortStep = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _prosemirrorModel = require("prosemirror-model");

var _prosemirrorTransform = require("prosemirror-transform");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var tableSortingStepType = 'atlaskit-table-sorting-ordering';
exports.tableSortingStepType = tableSortingStepType;

var TableSortStep = /*#__PURE__*/function (_Step) {
  (0, _inherits2.default)(TableSortStep, _Step);

  var _super = _createSuper(TableSortStep);

  function TableSortStep(pos, prev, next) {
    var _this;

    (0, _classCallCheck2.default)(this, TableSortStep);
    _this = _super.call(this);
    _this.prev = prev;
    _this.next = next;
    _this.pos = pos;
    return _this;
  }

  (0, _createClass2.default)(TableSortStep, [{
    key: "invert",
    value: function invert() {
      return new TableSortStep(this.pos, this.next, this.prev);
    }
  }, {
    key: "apply",
    value: function apply(doc) {
      return _prosemirrorTransform.StepResult.ok(doc);
    }
  }, {
    key: "map",
    value: function map() {
      return null;
    }
  }, {
    key: "getMap",
    value: function getMap() {
      return new _prosemirrorTransform.StepMap([0, 0, 0]);
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      return {
        stepType: tableSortingStepType
      };
    }
  }], [{
    key: "fromJSON",
    value: function fromJSON() {
      return new _prosemirrorTransform.ReplaceStep(0, 0, _prosemirrorModel.Slice.empty);
    }
  }]);
  return TableSortStep;
}(_prosemirrorTransform.Step);
/** Register this step with Prosemirror */


exports.TableSortStep = TableSortStep;

_prosemirrorTransform.Step.jsonID(tableSortingStepType, TableSortStep);
