// @atlaskit/adf-schema/dist/esm/next-schema/nodes/tableCell.js
import { adfNode } from '@atlaskit/adf-schema-generator';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
import { tableCellContentPseudoGroup } from '../groups/tableCellContentPseudoGroup';
export var tableCell = adfNode('tableCell').define({
  isolating: true,
  selectable: false,
  tableRole: 'cell',
  marks: [unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    colspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    rowspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    colwidth: {
      type: 'array',
      items: {
        type: 'number'
      },
      default: null,
      optional: true
    },
    background: {
      type: 'string',
      default: null,
      optional: true
    }
  },
  content: [tableCellContentPseudoGroup],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      required: {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: ['content']
      },
      'props.content': {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: 'tableCell_content'
      }
    }
  }
});
