// @atlaskit/adf-schema/dist/cjs/next-schema/full-schema.adf.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _blockGroup = require("./groups/blockGroup");
var _layoutSection = require("./nodes/layoutSection");
var _blockRootOnlyGroup = require("./groups/blockRootOnlyGroup");
var _blockContentGroup = require("./groups/blockContentGroup");
var _codeBlock = require("./nodes/codeBlock");
var _expand = require("./nodes/expand");
var doc = (0, _adfSchemaGenerator.adfNode)('doc').define({
  root: true,
  version: 1,
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_blockGroup.blockGroup, _blockContentGroup.blockContentGroup, _codeBlock.codeBlock.use('with_marks'), _layoutSection.layoutSection, _layoutSection.layoutSection.use('full'), _layoutSection.layoutSection.use('with_single_column'), _blockRootOnlyGroup.blockRootOnlyGroup, _expand.expand.use('with_breakout_mark')))],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.content.minItems': {
        remove: true,
        reason: "@DSLCompatibilityException - The content expression and minItems don't match in the current validator schema."
      }
    }
  }
});
var _default = exports.default = doc;
