// @atlaskit/adf-schema/dist/cjs/schema/nodes/extension.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.extension = void 0;

var _extensions = require("../../utils/extensions");

var createExtensionNodeSpec = function createExtensionNodeSpec() {
  var nodeSpec = {
    inline: false,
    group: 'block',
    atom: true,
    selectable: true,
    attrs: {
      extensionType: {
        default: ''
      },
      extensionKey: {
        default: ''
      },
      parameters: {
        default: null
      },
      text: {
        default: null
      },
      layout: {
        default: 'default'
      },
      localId: {
        default: null
      }
    },
    parseDOM: [{
      tag: '[data-node-type="extension"]',
      getAttrs: function getAttrs(domNode) {
        return (0, _extensions.getExtensionAttrs)(domNode);
      }
    }],
    toDOM: function toDOM(node) {
      var attrs = {
        'data-node-type': 'extension',
        'data-extension-type': node.attrs.extensionType,
        'data-extension-key': node.attrs.extensionKey,
        'data-text': node.attrs.text,
        'data-parameters': JSON.stringify(node.attrs.parameters),
        'data-layout': node.attrs.layout,
        'data-local-id:': node.attrs.localId
      };
      return ['div', attrs];
    }
  };
  return nodeSpec;
};

var extension = createExtensionNodeSpec();
exports.extension = extension;
