// @atlaskit/adf-schema/dist/cjs/schema/nodes/inline-extension.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inlineExtension = void 0;

var _extensions = require("../../utils/extensions");

var createInlineExtensionNodeSpec = function createInlineExtensionNodeSpec() {
  var nodeSpec = {
    inline: true,
    group: 'inline',
    selectable: true,
    attrs: {
      extensionType: {
        default: ''
      },
      extensionKey: {
        default: ''
      },
      parameters: {
        default: null
      },
      text: {
        default: null
      },
      localId: {
        default: null
      }
    },
    parseDOM: [{
      tag: 'span[data-extension-type]',
      getAttrs: function getAttrs(domNode) {
        return (0, _extensions.getExtensionAttrs)(domNode, true);
      }
    }],
    toDOM: function toDOM(node) {
      var attrs = {
        'data-extension-type': node.attrs.extensionType,
        'data-extension-key': node.attrs.extensionKey,
        'data-text': node.attrs.text,
        'data-parameters': JSON.stringify(node.attrs.parameters),
        'data-local-id:': node.attrs.localId,
        contenteditable: 'false'
      };
      return ['span', attrs];
    }
  };
  return nodeSpec;
};

var inlineExtension = createInlineExtensionNodeSpec();
exports.inlineExtension = inlineExtension;
