// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/table.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.table = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _fragment = require("../marks/fragment");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _tableRow = require("./tableRow");
var table = exports.table = (0, _adfSchemaGenerator.adfNode)('table').define({
  isolating: true,
  selectable: true,
  tableRole: 'table',
  marks: [_fragment.fragment, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    displayMode: {
      type: 'enum',
      values: ['default', 'fixed'],
      default: null,
      optional: true
    },
    isNumberColumnEnabled: {
      type: 'boolean',
      default: false,
      optional: true
    },
    layout: {
      type: 'enum',
      values: ['wide', 'full-width', 'center', 'align-end', 'align-start', 'default'],
      default: 'default',
      optional: true
    },
    localId: {
      type: 'string',
      minLength: 1,
      default: '',
      optional: true
    },
    width: {
      type: 'number',
      default: null,
      optional: true
    },
    __autoSize: {
      type: 'boolean',
      default: false,
      optional: true
    }
  },
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_tableRow.tableRow))]
});
