// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/tableRow.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tableRow = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _tableCell = require("./tableCell");
var _tableHeader = require("./tableHeader");
var _tableCellContent = require("./tableCellContent");
var tableRow = exports.tableRow = (0, _adfSchemaGenerator.adfNode)('tableRow').define({
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [(0, _adfSchemaGenerator.$onePlus)((0, _adfSchemaGenerator.$or)(_tableCell.tableCell, _tableHeader.tableHeader, _tableCellContent.tableCellContent))],
  tableRole: 'row',
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.content.minItems': {
        reason: '@DSLCompatibilityException - required for tableRow validator spec',
        remove: true
      }
    }
  }
});
