// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/tableCell.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.tableCell = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var _tableCellContentPseudoGroup = require("../groups/tableCellContentPseudoGroup");
var tableCell = exports.tableCell = (0, _adfSchemaGenerator.adfNode)('tableCell').define({
  isolating: true,
  selectable: false,
  tableRole: 'cell',
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  attrs: {
    colspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    rowspan: {
      type: 'number',
      default: 1,
      optional: true
    },
    colwidth: {
      type: 'array',
      items: {
        type: 'number'
      },
      default: null,
      optional: true
    },
    background: {
      type: 'string',
      default: null,
      optional: true
    }
  },
  content: [_tableCellContentPseudoGroup.tableCellContentPseudoGroup],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      required: {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: ['content']
      },
      'props.content': {
        reason: '@DSLCompatibilityException - required for tableHeader validator spec',
        value: 'tableCell_content'
      }
    }
  }
});
