// @atlaskit/adf-schema/dist/es2019/schema/nodes/extension.js
import { getExtensionAttrs } from '../../utils/extensions';

const createExtensionNodeSpec = () => {
  const nodeSpec = {
    inline: false,
    group: 'block',
    atom: true,
    selectable: true,
    attrs: {
      extensionType: {
        default: ''
      },
      extensionKey: {
        default: ''
      },
      parameters: {
        default: null
      },
      text: {
        default: null
      },
      layout: {
        default: 'default'
      },
      localId: {
        default: null
      }
    },
    parseDOM: [{
      tag: '[data-node-type="extension"]',
      getAttrs: domNode => getExtensionAttrs(domNode)
    }],

    toDOM(node) {
      const attrs = {
        'data-node-type': 'extension',
        'data-extension-type': node.attrs.extensionType,
        'data-extension-key': node.attrs.extensionKey,
        'data-text': node.attrs.text,
        'data-parameters': JSON.stringify(node.attrs.parameters),
        'data-layout': node.attrs.layout,
        'data-local-id:': node.attrs.localId
      };
      return ['div', attrs];
    }

  };
  return nodeSpec;
};

export const extension = createExtensionNodeSpec();
