// @atlaskit/adf-schema/dist/es2019/schema/nodes/inline-extension.js
import { getExtensionAttrs } from '../../utils/extensions';

const createInlineExtensionNodeSpec = () => {
  const nodeSpec = {
    inline: true,
    group: 'inline',
    selectable: true,
    attrs: {
      extensionType: {
        default: ''
      },
      extensionKey: {
        default: ''
      },
      parameters: {
        default: null
      },
      text: {
        default: null
      },
      localId: {
        default: null
      }
    },
    parseDOM: [{
      tag: 'span[data-extension-type]',
      getAttrs: domNode => getExtensionAttrs(domNode, true)
    }],

    toDOM(node) {
      const attrs = {
        'data-extension-type': node.attrs.extensionType,
        'data-extension-key': node.attrs.extensionKey,
        'data-text': node.attrs.text,
        'data-parameters': JSON.stringify(node.attrs.parameters),
        'data-local-id:': node.attrs.localId,
        contenteditable: 'false'
      };
      return ['span', attrs];
    }

  };
  return nodeSpec;
};

export const inlineExtension = createInlineExtensionNodeSpec();
