// @atlaskit/adf-schema/dist/es2019/schema/nodes/paragraph.js
/**
 * @name paragraph_node
 */

/**
 * @name paragraph_with_no_marks_node
 */

/**
 * NOTE: Need this because TS is too smart and inline everything.
 * So we need to give them separate identity.
 * Probably there's a way to solve it but that will need time and exploration.
 * // http://bit.ly/2raXFX5
 * type T1 = X | Y
 * type T2 = A | T1 | B // T2 = A | X | Y | B
 */

/**
 * @name paragraph_with_alignment_node
 */

/**
 * @name paragraph_with_indentation_node
 */
const isImageNode = node => {
  return Boolean(node && node.nodeName.toLowerCase() === 'img');
};

const hasInlineImage = node => {
  if (!node) {
    return false;
  }

  return Array.from(node.childNodes).some(child => {
    const isImage = isImageNode(child);

    if (!isImage && child.childNodes) {
      return Array.from(node.childNodes).some(node => hasInlineImage(node));
    }

    return isImage;
  });
};

const pDOM = ['p', 0];
export const paragraph = {
  selectable: false,
  content: 'inline*',
  group: 'block',
  marks: 'strong code em link strike subsup textColor typeAheadQuery underline confluenceInlineComment action annotation unsupportedMark unsupportedNodeAttribute dataConsumer fragment',
  parseDOM: [{
    tag: 'p'
  }],

  toDOM() {
    return pDOM;
  }

};
