// @atlaskit/adf-schema/dist/cjs/steps/analytics.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.analyticsStepType = exports.analyticsInvertStepType = exports.AnalyticsStep = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _prosemirrorTransform = require("prosemirror-transform");

var _prosemirrorModel = require("prosemirror-model");

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var analyticsStepType = 'atlaskit-analytics';
exports.analyticsStepType = analyticsStepType;
var analyticsInvertStepType = 'atlaskit-analytics-invert';
exports.analyticsInvertStepType = analyticsInvertStepType;
var HISTORY_ACTIONS;

(function (HISTORY_ACTIONS) {
  HISTORY_ACTIONS["UNDID"] = "undid";
  HISTORY_ACTIONS["REDID"] = "redid";
})(HISTORY_ACTIONS || (HISTORY_ACTIONS = {}));

/** Creates undo event from a normal analytics event */
function createUndoEvent(analyticsEvent) {
  var _analyticsEvent$paylo;

  return _objectSpread(_objectSpread({}, analyticsEvent), {}, {
    payload: {
      action: HISTORY_ACTIONS.UNDID,
      actionSubject: analyticsEvent.payload.actionSubject,
      actionSubjectId: analyticsEvent.payload.action,
      attributes: _objectSpread(_objectSpread({}, analyticsEvent.payload.attributes), {}, {
        actionSubjectId: analyticsEvent.payload.actionSubjectId,
        inputMethod: ((_analyticsEvent$paylo = analyticsEvent.payload.attributes) === null || _analyticsEvent$paylo === void 0 ? void 0 : _analyticsEvent$paylo.inputMethod) || ''
      }),
      eventType: 'track'
    }
  });
}
/** Toggles event action between undo & redo */


var toggleEventAction = function toggleEventAction(analyticsEvent) {
  return _objectSpread(_objectSpread({}, analyticsEvent), {}, {
    payload: _objectSpread(_objectSpread({}, analyticsEvent.payload), {}, {
      action: analyticsEvent.payload.action === HISTORY_ACTIONS.UNDID ? HISTORY_ACTIONS.REDID : HISTORY_ACTIONS.UNDID
    })
  });
};

function isHistoryAnalyticsEvent(event) {
  return event.payload.action === HISTORY_ACTIONS.UNDID || event.payload.action === HISTORY_ACTIONS.REDID;
}

/**
 * Custom Prosemirror Step to fire our GAS V3 analytics events
 * Using a Step means that it will work with prosemirror-history and we get
 * undo/redo events for free
 */
var AnalyticsStep = /*#__PURE__*/function (_Step) {
  (0, _inherits2.default)(AnalyticsStep, _Step);

  var _super = _createSuper(AnalyticsStep);

  function AnalyticsStep(analyticsEvents) {
    var _this;

    var actionsToIgnore = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
    var pos // Used to create the map, prevent splitting history.
    = arguments.length > 2 ? arguments[2] : undefined;
    (0, _classCallCheck2.default)(this, AnalyticsStep);
    _this = _super.call(this);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "analyticsEvents", []);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "actionsToIgnore", []);
    _this.analyticsEvents = analyticsEvents;
    _this.actionsToIgnore = actionsToIgnore;
    _this.pos = pos;
    return _this;
  }
  /**
   * Generate new undo/redo analytics event when step is inverted
   */


  (0, _createClass2.default)(AnalyticsStep, [{
    key: "invert",
    value: function invert() {
      var _this2 = this;

      var analyticsEvents = this.analyticsEvents.filter(function (analyticsEvent) {
        return _this2.actionsToIgnore.indexOf(analyticsEvent.payload.action) === -1;
      }).map(function (analyticsEvent) {
        if (isHistoryAnalyticsEvent(analyticsEvent)) {
          return toggleEventAction(analyticsEvent);
        } else {
          return createUndoEvent(analyticsEvent);
        }
      });
      return new AnalyticsStep(analyticsEvents, []);
    }
  }, {
    key: "apply",
    value: function apply(doc) {
      return _prosemirrorTransform.StepResult.ok(doc);
    }
  }, {
    key: "map",
    value: function map(mapping) {
      var newPos = this.pos;

      if (typeof newPos === 'number') {
        newPos = mapping.map(newPos);
      } // Return the same events, this step will never be removed


      return new AnalyticsStep(this.analyticsEvents, this.actionsToIgnore, newPos);
    }
  }, {
    key: "getMap",
    value: function getMap() {
      if (typeof this.pos === 'number') {
        return new _prosemirrorTransform.StepMap([this.pos, 0, 0]);
      }

      return new _prosemirrorTransform.StepMap([]);
    }
  }, {
    key: "merge",
    value: function merge(other) {
      if (other instanceof AnalyticsStep) {
        var otherAnalyticsEvents = other.analyticsEvents;
        return new AnalyticsStep([].concat((0, _toConsumableArray2.default)(otherAnalyticsEvents), (0, _toConsumableArray2.default)(this.analyticsEvents)));
      }

      return null;
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      return {
        stepType: analyticsStepType
      };
    }
  }], [{
    key: "fromJSON",
    value: function fromJSON() {
      return new _prosemirrorTransform.ReplaceStep(0, 0, _prosemirrorModel.Slice.empty);
    }
  }]);
  return AnalyticsStep;
}(_prosemirrorTransform.Step);
/** Register this step with Prosemirror */


exports.AnalyticsStep = AnalyticsStep;

_prosemirrorTransform.Step.jsonID(analyticsStepType, AnalyticsStep);
