// @atlaskit/adf-schema/dist/es2019/next-schema/nodes/table.js
import { $onePlus, $or, adfNode } from '@atlaskit/adf-schema-generator';
import { fragment } from '../marks/fragment';
import { unsupportedMark } from '../marks/unsupportedMark';
import { unsupportedNodeAttribute } from '../marks/unsupportedNodeAttribute';
import { tableRow } from './tableRow';
export const table = adfNode('table').define({
  isolating: true,
  selectable: true,
  tableRole: 'table',
  marks: [fragment, unsupportedMark, unsupportedNodeAttribute],
  attrs: {
    displayMode: {
      type: 'enum',
      values: ['default', 'fixed'],
      default: null,
      optional: true
    },
    isNumberColumnEnabled: {
      type: 'boolean',
      default: false,
      optional: true
    },
    layout: {
      type: 'enum',
      values: ['wide', 'full-width', 'center', 'align-end', 'align-start', 'default'],
      default: 'default',
      optional: true
    },
    localId: {
      type: 'string',
      minLength: 1,
      default: '',
      optional: true
    },
    width: {
      type: 'number',
      default: null,
      optional: true
    },
    __autoSize: {
      type: 'boolean',
      default: false,
      optional: true
    }
  },
  content: [$onePlus($or(tableRow))]
});
