// @atlaskit/adf-schema/dist/cjs/next-schema/nodes/paragraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.paragraph = void 0;
var _adfSchemaGenerator = require("@atlaskit/adf-schema-generator");
var _atomicInlineGroup = require("../groups/atomicInlineGroup");
var _inlineContentGroup = require("../groups/inlineContentGroup");
var _inlineGroup = require("../groups/inlineGroup");
var _alignmentAndIndentation = require("../marks/alignmentAndIndentation");
var _unsupportedMark = require("../marks/unsupportedMark");
var _unsupportedNodeAttribute = require("../marks/unsupportedNodeAttribute");
var paragraph = exports.paragraph = (0, _adfSchemaGenerator.adfNode)('paragraph').define({
  attrs: {
    localId: {
      type: 'string',
      optional: true,
      default: null
    }
  },
  selectable: false,
  marks: [_unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  hasEmptyMarks: true,
  content: [(0, _adfSchemaGenerator.$zeroPlus)((0, _adfSchemaGenerator.$or)(_inlineGroup.inlineGroup, _atomicInlineGroup.atomicInlineGroup, _inlineContentGroup.inlineContentGroup))],
  DANGEROUS_MANUAL_OVERRIDE: {
    'validator-spec': {
      'props.marks': {
        value: {
          type: 'array',
          items: [],
          optional: true
        },
        reason: 'NoMark type used in old schema had a hack maxItems: 0 to mark nodes that should not have any marks. This is not possible to represent in new DSL.'
      }
    }
  }
}).variant('with_alignment', {
  marks: [_alignmentAndIndentation.alignment, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [],
  ignore: []
}).variant('with_indentation', {
  marks: [_alignmentAndIndentation.indentation, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [],
  ignore: []
}).variant('with_marks', {
  marks: [_alignmentAndIndentation.alignment, _unsupportedMark.unsupportedMark, _unsupportedNodeAttribute.unsupportedNodeAttribute],
  content: [],
  ignore: []
}).variant('with_no_marks', {
  content: [],
  marks: [],
  noMarks: true,
  ignore: []
});
